### PR TITLE
cpu/stm32-common: add support for lpuart

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds support for LPUART to STM32 CPUs.
The current implementation is highly inspired from the kinetis one. It's possible to choose a specific clock to use the LPUART. LSE clock will allow the UART to run while in deep sleep mode but I didn't test this.

The nucleo-l433rc configuration was updated to have the stdio available via st-link (USB) and the nucleo-l073 configuration was also updated with the third UART mapped to LPUART1.

Looking at STM32 cpus that we support, only LPUART1 is available. That's why some part of the implementation are specific to this.

### Issues/PRs references

None

### CODE REVIEW

**Comment by aabadie** [comment]
This PR is ready for review in his actual state. What's done and tested:
- support for LPUART L4 and L0 works
- tested on third uart of nucleo-l073
- tested on first UART (STDIO) for nucleo-l433rc
- tested different clock sources (including LSE) with success, but needs some tweak to set the STDIO baudrate at the right speed
- adapted all L0/L4 based boards UART config

What's missing (or could be added alter):
- adapt nucleo-l496zg STDIO UART to use LPUART: it doesn't work for the moment, although everything seems valid
- implement multi clock sources for USART as well. This will only be available on L0/L4 series.

**Comment by aabadie** [comment]
Is there someone interested by this one ? It was tested on all STM32 boards but nucleo-l452. Nucleo-l496 doesn't work and I don't know why. I can remove the change related to this board to speed-up the merge.

**Comment by kYc0o** [comment]
> I can remove the change related to this board to speed-up the merge.

+1 👍

**Comment by aabadie** [comment]
Rebased.

> I can remove the change related to this board to speed-up the merge.

Well, in fact it's already removed. But all L0/L4 based boards need adaption in their UART periph conf.

File: boards/nucleo-l433rc/Makefile.include [hunk 1]
```diff
@@ -4,7 +4,7 @@ export CPU_MODEL = stm32l433rc
 
 # stdio is not available over st-link but on the Arduino TX/RX pins
 # A serial to USB converter plugged to the host is required
-PORT_LINUX ?= /dev/ttyUSB0
+PORT_LINUX ?= /dev/ttyACM0
```
Comment (kYc0o) [inline boards/nucleo-l433rc/Makefile.include]: Maybe you can remove the comment above?

Comment (aabadie) [reply boards/nucleo-l433rc/Makefile.include]: Of course +1

Comment (aabadie) [reply boards/nucleo-l433rc/Makefile.include]: Even better: use the common configuration provided for nucleo64 boards ;)

File: boards/nucleo-l073rz/include/periph_conf.h [hunk 2]
```diff
@@ -95,23 +97,27 @@ static const uart_conf_t uart_config[] = {
...
         .tx_pin     = GPIO_PIN(PORT_C, 10),
-        .rx_af      = GPIO_AF6,
-        .tx_af      = GPIO_AF6,
+        .rx_af      = GPIO_AF0,
+        .tx_af      = GPIO_AF0,
         .bus        = APB1,
-        .irqn       = USART4_5_IRQn
+        .irqn       = LPUART1_IRQn,
+        .type       = STM32_LPUART,
+        .clk_src    = 0, /* Use APB clock */
```
Comment (kYc0o) [inline boards/nucleo-l073rz/include/periph_conf.h]: This change forces the driver to use lpuart, which is compiled and thus takes a bit more space than how it is in current master, e.g. with tests/periph_uart with this PR:

```console
  text	   data	    bss	    dec	    hex	filename
  15580	    504	   4752	  20836	   5164	/Users/facosta/git/RIOT-OS/RIOT/tests/periph_uart/bin/nucleo-l073rz/tests_periph_uart.elf
```

and with master:

```console
  text	   data	    bss	    dec	    hex	filename
  14892	    504	   4752	  20148	   4eb4	/Users/facosta/git/RIOT-OS/RIOT/tests/periph_uart/bin/nucleo-l073rz/tests_periph_uart.elf
```

With this PR on examples/hello-world:

```console
   text	   data	    bss	    dec	    hex	filename
   9252	    140	   2748	  12140	   2f6c	/Users/facosta/git/RIOT-OS/RIOT/examples/hello-world/bin/nucleo-l073rz/hello-world.elf
```

In current master:

```console
   text	   data	    bss	    dec	    hex	filename
   8564	    140	   2748	  11452	   2cbc	/Users/facosta/git/RIOT-OS/RIOT/examples/hello-world/bin/nucleo-l073rz/hello-world.elf
```

Thus, I'd vote to remove it but then it comes the problem on how to show/advertise the support of this feature? Can it be better a submodule of periph_uart?

Comment (aabadie) [reply boards/nucleo-l073rz/include/periph_conf.h]: Your results make sense since the configuration of the nucleo-l073 board provides support for the 2 types of UARTs and thus more functions are built and included in the resulting firmware. But the ~700 bytes addition seems quite a lot and this seems weird, especially with the hello-world example. I'll look at it again.

Since this PR uses the same strategy implemented for the kinetis (the 2 types of UART drivers are implemented in the same file but one type can be skipped if not configured by the board), I would be interested to have the results of the build for one of them.

> Thus, I'd vote to remove it but then it comes the problem on how to show/advertise the support of this feature?

In fact, the nucleo-L073rz is the only L0 based board that is configured with an LPUART. This is not only useful for the given example but also for the CI, since it ensures the LPUART implementation is built at least for one L0 based board.

Comment (aabadie) [reply boards/nucleo-l073rz/include/periph_conf.h]: > it ensures the LPUART implementation is built at least for one L0 based board.

Looking again at the changes, the lpuart functions were built for all L0/L4 based boards. I pushed an update to limit this to the boards that actually configure an LPUART, using a define.
Fixing the size increase with hello-world would be possible but would require a lot more work and is out of the scope of this PR: it would require the application to declare which UART is needed and then use some conditional macro to include or not the requested UART.

Comment (kYc0o) [reply boards/nucleo-l073rz/include/periph_conf.h]: Then the question remains: is it valid enough to increase our sizes by 700B for the sake of LPUART?

@bergzand is your tool to compare sizes between master and PRs still working?

Comment (aabadie) [reply boards/nucleo-l073rz/include/periph_conf.h]: > is it valid enough to increase our sizes by 700B for the sake of LPUART?

From a CI point of view, I would say yes because now only this board provide an LPUART interface for L0. Having support for LPUART is also a nice feature for low-power applications.

**Comment by aabadie** [comment]
@kYc0o, how do we proceed here. I don't see more things that I can do until we have more fine grained configuration available at application level.

**Comment by kYc0o** [comment]
I'd say we need first to provide this granularity, since the code size increases considerably. #8711 would help a lot in this, since there's more or less the same problem to activate `ld -r`.

**Comment by aabadie** [comment]
>  we need first to provide this granularity, since the code size increases considerably

Are you taking into account that this is the case by default for only one board (nucleo-l073) that provides both USART and LPUART ?
Because otherwise I fear that this PR will become stuck again for several months (not to say years) until we have a configuration tool that provides the required granularity (if it does).

**Comment by aabadie** [comment]
Ok, to move forward on this PR, I applied @cladmi's proposition to use the MODULE_PERIPH_LPUART macro when FEATURE_REQUIRED make variable is used in a Makefile. And it works ! Thanks a lot @cladmi for this :)

So an extra LPUART interface is now only added to nucleo-l073rz when this macro is defined. For Nucleo-l433rc, this feature is always required since the LPUART interface is used for stdio (USB port with st-link).

In the `tests/periph_uart` application Makefile, I added FEATURE_OPTIONAL += periph_lpuart so only boards providing this are concerned.

In the end, in terms of code size, building hello-world for nucleo-l073rz increases by16B (instead of the previous 700B). The 700B are only added when building periph_uart test application.

I tested multiple UART interfaces on nucleo-l073rz and nucleo-l0433rc with success.

So this is ready for review again.

**Comment by aabadie** [comment]
is there someone interested by lpuart on stm32s ?

File: cpu/stm32_common/periph/uart.c [hunk 3]
```diff
@@ -44,19 +46,16 @@ static inline USART_TypeDef *dev(uart_t uart)
...
-
-    assert(uart < UART_NUMOF);
-
-    /* save ISR context */
-    isr_ctx[uart].rx_cb = rx_cb;
-    isr_ctx[uart].arg   = arg;
+static inline void uart_init_usart(uart_t uart, uint32_t baudrate);
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L4)
+#ifdef MODULE_PERIPH_LPUART
+static inline void uart_init_lpuart(uart_t uart, uint32_t baudrate);
```
Comment (kYc0o) [inline cpu/stm32_common/periph/uart.c]: Any reason for this to be inline?

Comment (aabadie) [reply cpu/stm32_common/periph/uart.c]: Note that I took this from other piece of code in RIOT (widely used apparently), see stm32 i2c periph driver for example. I guess this is done for optimisation purpose: in theory the compiler replaces the call to an inline function by the actual code of the function. For init function, this is maybe not worth.
If I remove the `inline`, there's no difference in term of code size.

Comment (kYc0o) [reply cpu/stm32_common/periph/uart.c]: `inline` only gives a hint to the compiler to try to optimise the function, it might be ignored here. However, AFAIK, `static inline` should only be used on headers and not C files.

Comment (aabadie) [reply cpu/stm32_common/periph/uart.c]: > However, AFAIK, static inline should only be used on headers and not C files.

Can you provide a reference for this? Because if it's the case, there's a lot to change in RIOT's code ;)

```
$ git grep "static inline" *.c
525
```

Comment (kYc0o) [reply cpu/stm32_common/periph/uart.c]: I was wrong saying it "should". It is possible but I don't see here the benefits of doing it, since the functions are maybe too big to be inlined (probably the `inline` is ignored by the compiler), as you stated.

For the `inline` doc I was reading this: https://www.greenend.org.uk/rjk/tech/inline.html 

> For init function, this is maybe not worth.
If I remove the inline, there's no difference in term of code size.

I think it would be more correct if you do so.

Comment (aabadie) [reply cpu/stm32_common/periph/uart.c]: I disagree, using static inline is like having a macro. It is perfectly fine as it is now.

**Comment by aabadie** [comment]
I'd like to squash and merge this one. @kYc0o, I don't think your comment about `static inline` is worth blocking this PR. If you tried this PR already, please ACK and we move forward.

File: cpu/stm32_common/periph/uart.c (hunk 3 above)
Comment (kYc0o) [reply cpu/stm32_common/periph/uart.c]: I think we need a third advice here, since I'm not 100% convinced about using the `static inline` at this level.

Any opinions @cladmi @kaspar030 @miri64 @vincent-d ?

Comment (miri64) [reply cpu/stm32_common/periph/uart.c]: I think the compiler would treat them equally anyway.

Comment (aabadie) [reply cpu/stm32_common/periph/uart.c]: > I think the compiler would treat them equally anyway.

I agree and this change request is completely pointless. Can we move forward one day and merge this PR ?

**Comment by aabadie** [comment]
monthly ping

File: cpu/stm32_common/periph/uart.c (hunk 3 above)
Comment (kYc0o) [reply cpu/stm32_common/periph/uart.c]: > I think the compiler would treat them equally anyway.

Yes, but the discussion here is about readability and code style. I won't say this is violating something, just want to be sure that the reasons why the function is declared that way actually exist, and is not just a "it's done this way somewhere else" or "it's perfectly fine".

The change is actually unrelated to the purpose of this PR. If you want to change it to `static inline`, please open another PR to do so. However, for such a PR I won't have the time to review it.

Comment (kaspar030) [reply cpu/stm32_common/periph/uart.c]: As the compiler is free to ignore "inline" and will inline anyways, inline-wise (code-size wise) there's no difference.

A static function that's not "inline" will generate a warning, though, if unused. Thus in headers, functions need to be "static inline" as the function might not be used by every includer.

In ```*.c``` files, without "inline" there'd also be a warning if unused, which might be desirable. Other than that, I personally use "static inline" like @aabadie mentioned: instead of a macro. To me it's a message to the code reader: "this function is super small and I expect it to be inlined".

TL;DR the inline is fine here, the change request should at most be postponed into a general issue.

Comment (kYc0o) [reply cpu/stm32_common/periph/uart.c]: Thanks for the explanation @kaspar030! Would you mind to ack the code style? I can ACK the rest and merge it if other people don't have other requests.

**Comment by kYc0o** [comment]
Please squash.

**Comment by aabadie** [comment]
There are tons of 
```
/tmp/dwq.0.12608236929051253/587e9a0ffb65c685ee0c6248d87111e0/boards/common/nucleo/board.c:50:1: fatal error: error closing /tmp/ccen8abt.s: No space left on device
```
unrelated errors in Murdock. Is there any problem with the CI ?

**Comment by kaspar030** [comment]
> Is there any problem with the CI ?

seems like. I'll check.

**Comment by aabadie** [comment]
Finally all green! Thanks @kaspar030 for unlocking the situation.

**Comment by aabadie** [comment]
@kYc0o, @kaspar030, any chance to merge this one (it's already green) ?

**Comment by miri64** [comment]
Since the build is already half a week old, let's rebuild. @kYc0o ACK'd so I'd say we merge once it is green again. :-)
